run-tests.php

```diff
@@ -1209,6 +1268,261 @@ function run_all_tests($test_files, $env, $redir_tested = null)
 	}
 }
 
+function run_all_tests_parallel($test_files, $env, $redir_tested) {
+    global $workers, $test_idx, $test_cnt, $test_results, $failed_tests_file, $result_tests_file, $PHP_FAILED_TESTS;
+
+    // The PHP binary running run-tests.php, and run-tests.php itself
+    // This PHP executable is *not* necessarily the same as the tested version
+    $thisPHP = PHP_BINARY;
+    $thisScript = __FILE__;
+
+    $workerProcs = [];
+    $workerStdins = [];
+    $workerStdouts = [];
+    $workerStderrs = [];
+
+    echo "====⚡️===========================================================⚡️====\n";
+    echo "====⚡️==== WELCOME TO THE FUTURE: run-tests PARALLEL EDITION ====⚡️====\n";
+    echo "====⚡️===========================================================⚡️====\n";
+
+    // Because some of the PHP test suite has not been written with
+    // parallel execution in mind, it is not safe to just run any two tests
+    // concurrently.
+    // Therefore, we divide the test set into directories and test multiple
+    // directories at once, but not multiple tests within them.
+
+    $testDirs = [];
+
+    foreach ($test_files as $file) {
+        $dirSeparator = strrpos($file, DIRECTORY_SEPARATOR);
+        if ($dirSeparator !== FALSE) {
+            $testDirs[substr($file, 0, $dirSeparator)][] = $file;
+        } else {
+            $testDirs[""][] = $file;
+        }
+    }
+
+    $testDirsToGo = array_values($testDirs);
+    // Sort test dirs so the biggest ones are handled first, so we spend less
+    // time waiting on workers tasked with very large dirs.
+    // This is an ascending sort because items are popped off the end.
+    // Thank you Rasmus for this idea :)
+    usort($testDirsToGo, 'count');
+    $testsInProgress = 0;
+
+    echo "Isolated ", count($testDirsToGo), " directories to be tested in parallel.\n";
+
+    echo "Spawning workers… ";
+    for ($i = 1; $i <= $workers; $i++) {
+        $proc = proc_open(
+            $thisPHP . ' ' . escapeshellarg($thisScript),
```

Comment:
This should probably set some ini options, such as [those](https://github.com/amphp/parallel/blob/0ef74413afd6370d3c348f719d527e3155694504/lib/Process/ChannelledProcess.php#L34-L38) to prevent unexpected messages on STDOUT.

Reply:
Yes, it would make sense to log to `stderr`.

That said, wouldn't writing a user error handler that sends errors as messages work? This is my current plan.